### PR TITLE
[netcore] Implement Type.GetType () overloads with assembly/type resolve delegates.

### PR DESCRIPTION
**Pull request by vargaz**
Instead of extending the runtime gettype implementation, port the type parsing code
to C#.

### CODE REVIEW

File: mcs/class/System.Private.CoreLib/System/Type.cs [hunk 1]
```diff
@@ -104,4 +117,386 @@ internal virtual string FormatTypeName (bool serialization)
...
+            foreach (char c in name) {
+                if (Array.IndexOf<char> (SPECIAL_CHARS, c) >= 0)
+                    sb.Append ('\\');
+                sb.Append (c);
+            }
+
+			return sb.ToString ();
+        }
+
+		class ParsedName {
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System/Type.cs]: Wouldn't be better to be a struct or ValueTuple?

File: mcs/class/System.Private.CoreLib/System/Type.cs [hunk 2]
```diff
@@ -104,4 +117,386 @@ internal virtual string FormatTypeName (bool serialization)
 			return !object.ReferenceEquals (left, right);
 		}
 	}
+
+	internal sealed class TypeNameParser
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System/Type.cs]: Make it static and move to separate file

File: mcs/class/System.Private.CoreLib/System/Type.cs [hunk 3]
```diff
@@ -104,4 +117,386 @@ internal virtual string FormatTypeName (bool serialization)
...
+                if (Array.IndexOf<char> (SPECIAL_CHARS, c) >= 0)
+                    sb.Append ('\\');
+                sb.Append (c);
+            }
+
+			return sb.ToString ();
+        }
+
+		class ParsedName {
+			public List<string> Names;
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System/Type.cs]: Use `ReadonlySpan<char>` instead

File: mcs/class/System.Private.CoreLib/System/Type.cs [hunk 4]
```diff
@@ -104,4 +117,386 @@ internal virtual string FormatTypeName (bool serialization)
...
+			public List<ParsedName> TypeArguments;
+			public List<int> Modifiers;
+			public string AssemblyName;
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System/Type.cs]: Use ReadonlySpan<char> instead

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 3 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: Those values are eventually passed to methods expecting a string, so a string needs to be created anyway.

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 2 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: Fixed.

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 1 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: We use null for error checking.

Comment (marek-safar) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: Use can still use nullable for that

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 3 above)
Comment (marek-safar) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: Not always, there is foreach conversion in EscapeTypeName

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 1 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: Thats doable, not sure its worth it, this is not really perf sensitive code.

File: mcs/class/System.Private.CoreLib/System/Type.cs (hunk 3 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Type.cs]: That code only runs if the string contains characters to be escaped, which is pretty rare.

**Comment by lambdageek** [comment]
We already have a managed type parser based on the C code: mcs/class/corlib/System/TypeSpec.cs.  I guess now we have 2.

**Comment by marek-safar** [comment]
@lambdageek maybe we could have custom ALC context and handle that in managed code as well for core

**Pull request merged**